### DISCUSSION TITLE
Add Required/Validation to Relational Fields

### DISCUSSION THREAD

**Discussion by botzill** [directus/directus#2985] [Archive]
Currently, if we have an M2M field:

<img width="874" alt="Screen Shot 2020-11-11 at 23 34 53" src="https://user-images.githubusercontent.com/4686040/98866987-7bff4580-2476-11eb-9977-f70cb8518d51.png">

I see does not have the required option. I guess we need to create a custom webhook action to validate that?

**Comment by benhaynes** [comment]
We have discussed ways to add relational validation (min items, max items, etc), but need to decide if that goes in the FIELD, the RELATIONSHIP, or the INTERFACE OPTIONS. That's the main thing we need to figure out before moving forward.

**Reply by Fusseldieb** [reply]
Or maybe add a new tab called "Validation" for all fields, they being Relational or not.

At first thought it would contain only 1 field called "Validation", but in the future this same tab could eventually contain a code-field to allow JSON that accepts keys like "gt", "lt", "eq", "regex", etc that could be stacked to fine-tune the validation and maybe a simple regex checker with a test list (with "Match"/"Don't match" tags after it).

Just giving some ideas.

**Reply by benhaynes** [reply]
Yeah! This is what we'll be doing for all normal validation, relational validation is just a bit trickier. I think all validation can be handled via the "JSON builder", we just need to think through how to validate relational _counts_, etc.

Some examples and ideas (not meant to **all** be applied at once):

```json
{
	"_contains": "Directus",
	"_regex": "[a-fA-F0-9]",
	"_lt": "$FIELD(field_name)",
	"_gte": "$NOW",
	"_in": [
		"published",
		"draft"
	],
	"_count": {
		"_gte": "3",
	}
}
```

Another option for validating the item count would be something like:

```json
{
	"value": {
		"_gte": "$NOW"
	},
	"count": {
		"_gte": "5"
	}
}
```

**Reply by rijkvanzanten** [reply]
This is an interesting approach:

```
	"_count": {
		"_gte": "3",
	}
```

Allowing to use any of the validation fields on the `_count` of the item sounds very useful to me. It would also be possible to use it  for JSON / CSV fields

**Reply by benhaynes** [reply]
Yeah, and by that logic we could also have a `_length` ... though this can maybe be accomplished with a RegEx too.

**Reply by Fusseldieb** [reply]
I'm a bit out of the loop, but would `"_count"` count the relational items and then only insert if gt, lt or eq a specified amount of items? Or am I getting this wrong?

**Reply by benhaynes** [reply]
That's what I'm thinking, yeah. Maybe it could also be used to count the root items within JSON and CSV values too?

**Reply by botzill** [reply]
@benhaynes Can you, please, ad some documentation about these validators and how I can use them? 

As I understand we can add that in Roles Field Validation?

**Reply by Fusseldieb** [reply]
Holy cow @benhaynes, this would allow a crazy amount of validation fine-tuning! I don't know for what I would need that, but maybe someone out there does. All in all, I like it 😋

Imo, maybe we should add only non-relational fields for now (like you said) and in a later minor version include `_count`, so it doesn't hinder development of more important stuff.

**Reply by benhaynes** [reply]
@botzill — we're working on Docs, but you can add tickets to things that are missing... just make sure you look to see what's there arleady:

https://github.com/directus/next/blob/main/docs/reference/filter-rules.md
https://github.com/directus/next/blob/main/docs/reference/item-objects.md

**Reply by botzill** [reply]
@benhaynes yes I saw those pages, I'm interested to know how to set these parts:

<img width="856" alt="Screen Shot 2020-11-13 at 09 38 29" src="https://user-images.githubusercontent.com/4686040/99041653-fffb1f80-2593-11eb-809e-98d2f74fdd6c.png">

I want to create custom validators for fieds. Any example would be great.

**Comment by benhaynes** [comment]
As I said, you use Filter Rules. If a `title` field needs to contain the word "Directus", then you would use this:

```js
{
	"title": {
		"_contains": "Directus"
	}
}
```

https://github.com/directus/next/blob/main/docs/reference/filter-rules.md
